tests/common/include/tests/common/tools/create_toy_geometry.hpp

```diff
@@ -5,353 +5,928 @@
-        int n_z_bins = binning.second;
-
-        // module positions
-        detray::dvector<point3> m_centers;
-        m_centers.reserve(n_phi_bins * n_z_bins);
-
-        // prep work
-        scalar pi{static_cast<scalar>(M_PI)};
-        scalar phi_step = scalar{2} * pi / (n_phi_bins);
-        scalar min_phi = -pi + scalar{0.5} * phi_step;
-        scalar z_start =
-            scalar{-0.5} * (n_z_bins - 1) * (scalar{2} * m_half_y - l_overlap);
-        scalar z_step = scalar{2} * std::abs(z_start) / (n_z_bins - 1);
-
-        // loop over the bins
-        for (size_t z_bin = 0; z_bin < size_t(n_z_bins); ++z_bin) {
-            // prepare z and r
-            scalar m_z = z_start + z_bin * z_step;
-            scalar m_r = (z_bin % 2) != 0u
-                             ? layer_r - scalar{0.5} * radial_stagger
-                             : layer_r + scalar{0.5} * radial_stagger;
-            for (size_t phiBin = 0; phiBin < size_t(n_phi_bins); ++phiBin) {
-                // calculate the current phi value
-                scalar m_phi = min_phi + phiBin * phi_step;
-                m_centers.push_back(
-                    point3{m_r * std::cos(m_phi), m_r * std::sin(m_phi), m_z});
-            }
-        }
+    /** source link */
+    const dindex inv_sf_finder = dindex_invalid;
+    /** Leaving world */
+    const dindex leaving_world = dindex_invalid;
 
-        // Create surfaces
-        for (auto& m_center : m_centers) {
-
-            // Build the transform
-            // The local phi
-            scalar m_phi = algebra::getter::phi(m_center);
-            // Local z axis is the normal vector
-            vector3 m_local_z{std::cos(m_phi + m_tilt_phi),
-                              std::sin(m_phi + m_tilt_phi), 0.};
-            // Local x axis the normal to local y,z
-            vector3 m_local_x{-std::sin(m_phi + m_tilt_phi),
-                              std::cos(m_phi + m_tilt_phi), 0.};
-
-            // add transform
-            transforms[detector_t::e_rectangle2].emplace_back(
-                ctx, m_center, m_local_z, m_local_x);
-
-            // add mask
-            masks.template add_mask<detector_t::e_rectangle2>(m_half_x,
-                                                              m_half_y);
-            masks.template group<detector_t::e_rectangle2>().back().links() = {
-                volume_id, dindex_invalid};
+    // create four barrel layers with modules
+    std::size_t n_brl_layers = 4;
+    // create three endcap layers with modules
+    std::size_t n_edc_layers = 3;
+
+    //
+    // barrel
+    //
+    const scalar brl_half_z = 500.;
+    const std::vector<scalar> brl_positions = {19., 32., 72., 116., 172.};
+    const std::vector<std::pair<scalar, scalar>> brl_lay_sizes = {
+        {0., 27.}, {27., 38.}, {64., 80.}, {108., 124.}, {164., 180.}};
+    const std::vector<std::pair<int, int>> brl_binning = {
+        {0., 0.}, {16, 14}, {32, 14}, {52, 14}, {78, 14}};
```

Comment:
Why are there five layers rather than four as you mentioned? What is the point of having {0,0} binning?

Reply:
The first one is actually the beampipe volume. I agree, its kind of hiding there...
And since it only has one surface inside, I didn't put a gap layer in between. Maybe thats not a good idea

Reply:
OK